src/Proto.Mailbox/NonBlockingBoundedMailbox.cs

```diff
@@ -0,0 +1,48 @@
+﻿// -----------------------------------------------------------------------
+//   <copyright file="NonBlockingBoundedMailbox.cs" company="Asynkron HB">
+//       Copyright (C) 2015-2017 Asynkron HB All rights reserved
+//   </copyright>
+// -----------------------------------------------------------------------
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace Proto.Mailbox
+{
+    public class NonBlockingBoundedMailbox : IMailboxQueue
+    {
+        public NonBlockingBoundedMailbox(int maxSize, Action<object> overflowAction, TimeSpan timeout)
+        {
+            _maxSize = maxSize;
+            _overflowAction = overflowAction;
+            _timeout = timeout;
+        }
+
+        private readonly ConcurrentQueue<object> _messages = new ConcurrentQueue<object>();
+        private readonly int _maxSize;
+        private readonly Action<object> _overflowAction;
+        private readonly TimeSpan _timeout;
+
+        public void Push(object message)
+        {
+            if (SpinWait.SpinUntil(() => _messages.Count < _maxSize, _timeout))
+            {
+                //this will be racy, but best effort is good enough..
+                _messages.Enqueue(message);
+            }
```

Comment:
I've been wondering if we need the SpinWait around this code - I thought a ConcurrentQueue allows multiple threads to Count and Enqueue? I'm not an expert in multithreading by any means but when I remove this SpinWait it seems to work pretty much the same as with it, only much faster (just messing around with a test that creates lots of Tasks that post to the mailbox)